src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,1062 @@
+            positions = create_position_ids_from_input_ids(input, self.padding_idx)
+        return super().forward(positions)
+
+
+def LayerNorm(normalized_shape, eps=1e-5, elementwise_affine=True):
+    if torch.cuda.is_available():
+        try:
+            from apex.normalization import FusedLayerNorm
+
+            return FusedLayerNorm(normalized_shape, eps, elementwise_affine)
+        except ImportError:
+            pass
+    return torch.nn.LayerNorm(normalized_shape, eps, elementwise_affine)
+
+
+def fill_with_neg_inf(t):
+    """FP16-compatible function that fills a input_ids with -inf."""
+    return t.float().fill_(float("-inf")).type_as(t)
+
+
+def _filter_out_falsey_values(tup) -> Tuple:
+    """Remove entries that are None or [] from an iterable."""
+    return tuple(x for x in tup if isinstance(x, torch.Tensor) or x)
+
+
+# Public API
+
+
+@add_start_docstrings(
+    "The bare BART Model model outputting raw hidden-states without any specific head on top.",
+    BART_START_DOCSTRING,
+    BART_INPUTS_DOCSTRING,
+)
+class BartModel(PretrainedBartModel):
+    """"""
+
+    def __init__(self, config: BartConfig):
+        super().__init__(config)
+        self.output_attentions = config.output_attentions
+        self.output_hidden_states = config.output_hidden_states
+
+        padding_idx, vocab_size = config.pad_token_id, config.vocab_size
+        self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
+
+        self.encoder = BartEncoder(config, self.shared)
+        self.decoder = BartDecoder(config, self.shared)
+
+        self.init_weights()
+
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+
+    def get_output_embeddings(self):
+        return _make_linear_from_emb(self.shared)
+
+    def forward(
+        self,
+        input_ids,
+        attention_mask=None,
+        decoder_input_ids=None,
+        encoder_outputs=None,  # type: Tuple
+        decoder_attention_mask=None,
+        cached_states=None,
+        **unused  # TODO(SS): does fairseq have an equivalent to token_type_ids?
+    ):
+        """
+        Args:
+            input_ids:
+            attention_mask: Ignore pad tokens in the input_ids
+            decoder_input_ids:
+            encoder_outputs: None or Tuple. If a tuple is passed, its first entry is assumed to be the features. If
+            none is passed, we call the encoder.
+            decoder_attention_mask:
+            **unused:
+
+        Returns:
+
+        """
+        # make masks if user doesn't supply
+        attention_mask, decoder_input_ids, decoder_attn_mask = prepare_bart_inputs(
+            self.config,
+            input_ids,
+            attention_mask=attention_mask,
+            decoder_input_ids=decoder_input_ids,
+            decoder_attn_mask=decoder_attention_mask,
+        )
+        assert decoder_input_ids is not None
+        if encoder_outputs is None:
```

Comment:
That's great. So for language generation the encoder_outputs are calculated only for the first decoding step and are then reused for the auto-regressive decoding, right?

Reply:
They will be. Not implemented yet, there is just a todo.